### PR TITLE
🧵 feat: Background Tool Calls for Agents & Model Specs

### PR DESCRIPTION
**Pull request by danny-avila**
## Summary

Adds **opt-in, poll-based background tool calls** for agents and model specs. The model can mark an eligible tool call with an injected `run_in_background: true` parameter; the host executor registers a task, **returns a synthetic handle immediately** (so the graph superstep resolves and the turn continues), runs the real tool as a detached promise, and stores the result in an in-process registry. The model retrieves progress/result later via a new **`check_background_task`** tool.

A backgrounded call and the poll call are both synchronous from the graph's point of view, so **no `@librechat/agents` change is required**.

## Opt-in (mirrors `deferred_tools`)

- **Admin gate:** new `AgentCapabilities.run_in_background` (in `endpoints.agents.capabilities`). **Off by default** (experimental).
- **Saved agents:** per-tool `tool_options[name].run_in_background`.
- **Model specs / ephemeral agents:** `TModelSpec.runInBackground` and `TEphemeralAgent.run_in_background` synthesize per-tool `tool_options` in `load.ts` / `added.ts`. Both paths converge at `initializeAgent`, so the injection + execution logic is written once.

When the capability is on and a tool is opted in, a `run_in_background` boolean is injected into that tool's schema and the `check_background_task` poll tool is registered for the run.

## How it works

1. `initializeAgent` runs `applyBackgroundToolCalls` after all built-in tool registration: injects the param into eligible opted-in tools and registers the poll tool.
2. In `createToolExecuteHandler` (`ON_TOOL_EXECUTE`), a backgrounded call registers a task and returns `{ background_task_id, status: "running", ... }` immediately; the real tool runs detached and its result/error lands in the registry.
3. `check_background_task` is host-special-cased (like `read_file`/`skill`) and reads the registry.

The "background active" gate in the executor is `configurable.toolRegistry.has(check_background_task)` — exactly the condition under which the model could have been shown the injected param.

## Registry

In-process, scoped per `user + conversation`, **idempotent by `toolCallId`** (safe across graph resume/replay — never double-dispatches), capped per conversation, and TTL-swept. The detached promise is `.catch`-guarded (no unhandled rejections).

## Scope / follow-ups (intentionally out of this PR)

- **Push notifications:** poll-only for now. The clean path (host detects completion → drives a fresh turn → injects via the SDK's existing `injectedMessages`) is a separate effort; queue/steer is not a prerequisite.
- **Subagents:** they run the SDK's *direct* execution path, which host-side interception can't reach, so they're excluded from eligibility (only the primary agent gets the capability). Backgrounding subagents needs an SDK-side change.
- **Excluded tools:** code-execution / file-authoring / `read_file` / `skill` / memory / `ask_user_question` / PTC / `tool_search` / handoffs — they break on a synthetic (artifact-less) result or run off the event-driven path.
- **Multi-replica:** the in-process registry is lost on restart and not shared across Redis replicas without sticky sessions; a Redis-backed store slots behind the same interface later.
- **Overlay/`mcp_all` ephemeral MCP servers:** tool names aren't known at synthesis time, so they aren't marked.
- No agent-builder UI toggle yet (admins configure via `librechat.yaml` / model spec; saved agents via `tool_options`).

## Testing

- `background.spec.ts` (25) — eligibility, `run_in_background` detect/strip, schema injection (non-mutating), `applyBackgroundToolCalls` gating, per-tool synthesis, registry (create/complete/fail/idempotency/capacity/scoping), and `check_background_task` output.
- `handlers.background.spec.ts` (3) — drives the **real** `createToolExecuteHandler`: returns a handle immediately, runs the tool exactly once detached (without the injected flag), poll returns the result; idempotent on replay; synchronous when not requested.
- `load.spec.ts` (+2) — ephemeral flag and model-spec `runInBackground` synthesize `tool_options` for eligible tools only.
- `tsc` + `eslint` clean across `packages/api` and `data-provider`; no regressions in existing validation/handlers/load suites.

@codex review

### CODE REVIEW

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/handlers.ts [hunk 1]
```diff
@@ -3217,8 +3227,114 @@ export function createToolExecuteHandler(options: ToolExecuteOptions): EventHand
...
+                  );
+                } else {
+                  const strippedArgs = stripRunInBackgroundArg(tc.args);
+                  void (async () => {
+                    try {
+                      const result = (await tool.invoke(
+                        normalizeToolInvokeArgs(strippedArgs, tool),
+                        {
+                          toolCall: { id: tc.id, stepId: tc.stepId, turn: tc.turn },
+                          configurable: mergedConfigurable,
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/handlers.ts]: Intentional scope for this PR, not a silent gap. Raw background tool calls are deliberately **same-run** here — matching LangChain deepagents, which backgrounds *subagents* (their own durable run/thread), not raw tool calls, for exactly this request-lifecycle reason. Since the detached invoke shares the run's abort signal, a tool still running at turn end is aborted and our catch marks the task `error` (it never hangs as `running`), so a later poll reports cleanly. Durable, request-detached execution that outlives the run is a tracked follow-up, modeled as background **subagents** over the existing checkpointer + GenerationJobManager substrate (a run/thread is the right durable unit; a raw tool call is not). Leaving this thread open as the follow-up marker.

**Comment by danny-avila** [comment]
**Codex round 1 addressed in `7a45465`.** Resolved inline threads deleted per this PR's workflow; recorded here for the audit trail:

- **Registry not carried for normal tools (P1):** the executor gate no longer relies on `toolRegistry` in `configurable` (only present for PTC/tool_search). `initializeAgent` now threads the exact injected tool names via `configurable.backgroundToolNames`; ordinary tools dispatch correctly and the injected arg is stripped.
- **Per-tool opt-in enforced at execution:** background dispatch gated on `backgroundToolSet.has(name)`, so an extra arg on a non-opted-in tool can't bypass the per-tool control.
- **Poll-tool name collision:** `check_background_task` interception is gated on the run actually enabling background, so a user tool of that name still runs.
- **Added-convo agents:** `backgroundToolsAvailable` is now forwarded to `processAddedConvo`.
- **Artifact drop:** `web_search`/`file_search` excluded from eligibility (their sources/citations are only processed by the foreground `toolEndCallback`).

Request-lifecycle detachment (tasks outliving the run) is intentionally deferred to a durable follow-up (background subagents over the checkpointer substrate) — see the remaining open thread.

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/handlers.ts [hunk 2]
```diff
@@ -3217,8 +3226,113 @@ export function createToolExecuteHandler(options: ToolExecuteOptions): EventHand
...
+                    backgroundUserId,
+                    backgroundConversationId,
+                    task.id,
+                    `Tool ${tc.name} not found`,
+                  );
+                } else {
+                  const strippedArgs = stripRunInBackgroundArg(tc.args);
+                  void (async () => {
+                    try {
+                      const result = (await tool.invoke(
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/handlers.ts]: Same intentional scope as the earlier request-lifecycle thread. This PR is same-run by design (deepagents backgrounds *subagents*, not raw tool calls, for exactly this reason): the value is parallelism across a multi-step turn, and if a tool is still running when the turn's cleanup aborts, our catch marks the task `error` rather than hanging. Surviving `completeJob()`/the request abort to outlive the turn is the durable follow-up — modeled as background **subagents** over the existing checkpointer + GenerationJobManager substrate (a run/thread is the right durable unit; detaching a raw tool also means owning its MCP-connection lifetime). Keeping this open as the follow-up marker.

**Comment by danny-avila** [comment]
**Codex round 2 addressed in `445bb71f8`.** Resolved threads deleted per this PR's workflow; recorded here:

- **Idempotency across turns:** provider tool-call ids repeat (e.g. `call_0`), so the dedupe key is now `runId::toolCallId` (+ turn) and orphaned mappings are swept — a later turn no longer collides with a retained task.
- **Artifacts preserved:** a backgrounded tool's artifact is now processed through the same `toolEndCallback` as the foreground path (guarded/best-effort), so images/files/citations aren't dropped.
- **Child agents:** `backgroundToolsAvailable` is forwarded to connected-agent discovery and subagent `processAgent` init, so a child agent's own event-driven tools behave like when it runs as primary.
- **Foreground flag strip (P3):** the injected `run_in_background` is stripped on foreground calls of background-capable tools (model may emit `false`) so strict MCP/action schemas don't reject.
- **List path context guard:** `check_background_task` without an id returns metadata only (`result_available`/`result_chars`), never full results; the full result is returned only for a specific id.

Request abort/lifecycle detachment (tasks outliving the turn) remains the intentional same-run boundary; durable execution is the tracked subagent follow-up — see the open thread.

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/handlers.ts [hunk 3]
```diff
@@ -3217,8 +3226,151 @@ export function createToolExecuteHandler(options: ToolExecuteOptions): EventHand
...
+                        normalizeToolInvokeArgs(strippedArgs, tool),
+                        {
+                          toolCall: { id: tc.id, stepId: tc.stepId, turn: tc.turn },
+                          configurable: mergedConfigurable,
+                          metadata,
+                        } as Record<string, unknown>,
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/handlers.ts]: Same same-run boundary as the other lifecycle threads, now specific to the MCP request-scoped connection store. Backgrounded MCP calls are meant to run *within* the turn (parallel to the model's other steps) and be polled before `cleanupMCPRequestContextForReq` runs at turn end; a call that outlives the turn is the durable follow-up, which is exactly where owning the MCP connection/auth lifetime off-request belongs (modeled as background subagents over the checkpointer substrate). Keeping this open as the follow-up marker.

File: packages/api/src/agents/handlers.ts [hunk 4]
```diff
@@ -3217,8 +3226,151 @@ export function createToolExecuteHandler(options: ToolExecuteOptions): EventHand
...
+                      /** Process artifacts (images, files, UI resources,
+                       *  citations) through the same callback as the foreground
+                       *  path so backgrounding an artifact-producing tool doesn't
+                       *  drop them. Best-effort: if the turn already returned and
+                       *  the stream is closed, persistence still runs and the SSE
+                       *  write no-ops. */
+                      if (toolEndCallback && result.artifact != null) {
+                        try {
+                          await toolEndCallback(
+                            {
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/handlers.ts]: Partly addressed in `59559c035`, partly scoped. The named case — image-generation built-ins — is now **excluded from eligibility** (dalle/flux/gemini_image_gen/image_gen_oai/image_edit_oai), so their files aren't silently dropped. For other tools that finish *within* the turn, the artifact is still processed best-effort through the same `toolEndCallback` while `artifactPromises` are being collected. Attaching an artifact to an **already-saved** turn is genuinely impossible (BaseClient awaits promises only at save), so that case is the durable follow-up, not something this same-run PR can fix — the poll result flags `hasArtifact` so the model isn't misled. Keeping this open as a follow-up marker.

**Comment by danny-avila** [comment]
**Codex round 3 addressed in `59559c035`.** Resolved threads deleted; recorded here:

- **Eager execution:** background-capable tool names are added to the run's eager-execution exclusion set, so a speculative eager dispatch can't launch an uncancelable detached task with partial/stale args.
- **Reserved poll-tool name:** a colliding user/MCP `check_background_task` is now shadowed by the host poll schema (with a warning) so the advertised schema matches the executor's interception.
- **Subagents:** reverted — pure-subagent (spawn-tool) configs no longer receive `backgroundToolsAvailable`, since their tools don't reach the host interceptor. Kept for primary/added/connected agents. Subagent background = the durable follow-up.
- **All routes:** `backgroundToolsAvailable` + `backgroundToolNames` are now threaded through the OpenAI-compatible and Responses agent routes (was chat-only).

Two facets of the same-run boundary remain open threads (intentional, → durable subagent follow-up): request/MCP-connection lifetime for tasks that outlive the turn, and attaching artifacts to an already-saved turn. Image-generation built-ins are excluded as a concrete safeguard.

**Comment by danny-avila** [comment]
@codex review

**Comment by danny-avila** [comment]
**Codex round 4 addressed in `5e1fb3502`** (all correctness edge cases, no scope items). Resolved threads deleted; recorded here:

- **Self-spawn subagents:** the parent's injected inputs are now sanitized (strip `run_in_background` + `check_background_task`) before reuse for self-spawn, so the isolated child doesn't advertise a background schema it can't honor. (SDK resolver keeps a provided `agentInputs` even with `self: true`.)
- **PTC visibility:** `check_background_task` is filtered out of `run_tools_with_code` tool definitions — host-only, not callable from generated code.
- **Stringified args:** `run_in_background` is now detected/stripped even when a provider delivers args as a JSON string, so it's honored (not silently foregrounded) and never leaks to strict object-schema tools.
- **Tools owning the param / non-object schemas:** injection is skipped for tools that already declare their own `run_in_background` and for string-input schemas (no input-contract rewrite); such tools aren't marked background-capable.

Only the two intentional durable-boundary threads remain open (request/MCP lifetime + after-save artifacts → durable subagent follow-up).

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/handlers.ts [hunk 5]
```diff
@@ -3217,8 +3226,151 @@ export function createToolExecuteHandler(options: ToolExecuteOptions): EventHand
...
+                  );
+                } else {
+                  const strippedArgs = stripRunInBackgroundArg(tc.args);
+                  void (async () => {
+                    try {
+                      const result = (await tool.invoke(
+                        normalizeToolInvokeArgs(strippedArgs, tool),
+                        {
+                          toolCall: { id: tc.id, stepId: tc.stepId, turn: tc.turn },
+                          configurable: mergedConfigurable,
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/handlers.ts]: Re-raise of the intentional same-run boundary (request/MCP-connection lifetime). Backgrounded MCP calls are meant to run within the turn and be polled before the response's finish/close cleanup; a task that outlives the turn — and owning the MCP connection/auth lifetime off-request — is the durable subagent follow-up (checkpointer + GenerationJobManager). Not re-fixing here by design.

File: packages/api/src/agents/handlers.ts [hunk 6]
```diff
@@ -3217,8 +3226,151 @@ export function createToolExecuteHandler(options: ToolExecuteOptions): EventHand
...
+                          metadata,
+                        } as Record<string, unknown>,
+                      )) as { content?: unknown; artifact?: unknown };
+                      /** Process artifacts (images, files, UI resources,
+                       *  citations) through the same callback as the foreground
+                       *  path so backgrounding an artifact-producing tool doesn't
+                       *  drop them. Best-effort: if the turn already returned and
+                       *  the stream is closed, persistence still runs and the SSE
+                       *  write no-ops. */
+                      if (toolEndCallback && result.artifact != null) {
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/handlers.ts]: Re-raise of the intentional after-save-artifact boundary. Within-turn completions still attach best-effort via toolEndCallback; artifact-first built-ins (image gen) are excluded; attaching to an already-saved turn is impossible (BaseClient awaits artifactPromises only at save), so that case is the durable follow-up. The poll result flags hasArtifact so the model isn't misled. Not re-fixing here by design.

**Comment by danny-avila** [comment]
**Codex round 5 addressed in `dac5208b3`.** Resolved threads deleted; recorded here:

- **Poll with stringified args:** `check_background_task` now parses string-delivered args, so providers that send a JSON string can retrieve a specific task by id (previously fell through to the metadata-only list).
- **Multi-agent dedupe:** the background dedupe key now includes `agentId` (`agentId::runId::toolCallId`), so two agents in one run emitting the same provider id (e.g. `call_0`) launch independent tasks instead of colliding.
- **Self-spawn registry:** self-spawn sanitization now also strips the background entries from the reused `toolRegistry` (not just `toolDefinitions`), so a child using tool_search/deferred loading can't rediscover the host-only background schema.

The same two intentional durable-boundary threads remain open (request/MCP lifetime + after-save artifacts → durable subagent follow-up).

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/handlers.ts [hunk 7]
```diff
@@ -3217,8 +3226,153 @@ export function createToolExecuteHandler(options: ToolExecuteOptions): EventHand
...
+                  );
+                } else {
+                  const strippedArgs = stripRunInBackgroundArg(tc.args);
+                  void (async () => {
+                    try {
+                      const result = (await tool.invoke(
+                        normalizeToolInvokeArgs(strippedArgs, tool),
+                        {
+                          toolCall: { id: tc.id, stepId: tc.stepId, turn: tc.turn },
+                          configurable: mergedConfigurable,
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/handlers.ts]: Re-raise of the decided same-run boundary (request/MCP-connection lifetime). Backgrounded MCP calls run within the turn and are polled before response cleanup; surviving cleanup to outlive the turn (a background-owned connection context) is the durable subagent follow-up. Not re-fixing here by design.

File: packages/api/src/agents/handlers.ts [hunk 8]
```diff
@@ -3217,8 +3226,153 @@ export function createToolExecuteHandler(options: ToolExecuteOptions): EventHand
...
+                          metadata,
+                        } as Record<string, unknown>,
+                      )) as { content?: unknown; artifact?: unknown };
+                      /** Process artifacts (images, files, UI resources,
+                       *  citations) through the same callback as the foreground
+                       *  path so backgrounding an artifact-producing tool doesn't
+                       *  drop them. Best-effort: if the turn already returned and
+                       *  the stream is closed, persistence still runs and the SSE
+                       *  write no-ops. */
+                      if (toolEndCallback && result.artifact != null) {
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/handlers.ts]: Re-raise of the decided after-save-artifact boundary. Within-turn completions attach best-effort; artifact-first built-ins (image gen) are excluded; a durable artifact path for tools that finish after the turn saves is the durable follow-up. The poll flags hasArtifact so the model isn't misled. Not re-fixing here by design.

**Comment by danny-avila** [comment]
**Codex round 6 addressed in `7c3b85fb9`.** One real fix (resolved thread deleted):

- **PTC target schemas:** the programmatic-tool-calling path already excluded the host-only `check_background_task`, but target tool schemas still advertised the injected `run_in_background` param (the shared registry entries were mutated). PTC codegen doesn't go through the host interceptor, so it's now stripped from PTC toolDefs like the self-spawn path.

The two intentional durable-boundary items (request/MCP lifetime + after-save artifacts) were re-raised and remain the durable subagent follow-up — replied, not re-fixed.

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/handlers.ts [hunk 9]
```diff
@@ -3217,8 +3227,153 @@ export function createToolExecuteHandler(options: ToolExecuteOptions): EventHand
...
+                        try {
+                          await toolEndCallback(
+                            {
+                              output: {
+                                name: tc.name,
+                                tool_call_id: tc.id,
+                                content: result.content,
+                                artifact: result.artifact,
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/handlers.ts]: Re-raise of the decided after-save-artifact boundary. Within-turn completions attach best-effort; artifact-first built-ins are excluded; a durable artifact path for tools finishing after the turn saves is the durable subagent follow-up. The poll flags hasArtifact so the model isn't misled. Not re-fixing here by design.

File: packages/api/src/agents/handlers.ts [hunk 10]
```diff
@@ -3217,8 +3227,153 @@ export function createToolExecuteHandler(options: ToolExecuteOptions): EventHand
...
+                  );
+                } else {
+                  const strippedArgs = stripRunInBackgroundArg(tc.args);
+                  void (async () => {
+                    try {
+                      const result = (await tool.invoke(
+                        normalizeToolInvokeArgs(strippedArgs, tool),
+                        {
+                          toolCall: { id: tc.id, stepId: tc.stepId, turn: tc.turn },
+                          configurable: mergedConfigurable,
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/handlers.ts]: Re-raise of the decided same-run/MCP-connection-lifetime boundary. A background-owned connection context that survives response cleanup is the durable subagent follow-up. Not re-fixing here by design.

**Comment by danny-avila** [comment]
**Codex round 7 addressed in `1177e6baa`.** One real fix (resolved thread deleted):

- **Explicit subagent inputs:** a child agent reachable as a top-level/handoff agent is initialized WITH the background capability, then reused as an explicit subagent. Round 4 only sanitized self-spawn; explicit child `agentInputs` are now sanitized the same way (`stripBackgroundFromToolDefinitions`/`Registry`) whenever `child.backgroundToolNames` is non-empty, so an isolated child graph doesn't advertise a background contract it can't honor.

The two intentional durable-boundary items (MCP connection lifetime + after-save artifacts) were re-raised and remain the durable subagent follow-up — replied, not re-fixed.

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/handlers.ts [hunk 11]
```diff
@@ -3217,8 +3227,153 @@ export function createToolExecuteHandler(options: ToolExecuteOptions): EventHand
...
+                          metadata,
+                        } as Record<string, unknown>,
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/handlers.ts]: Re-raise of the decided same-run/request-abort boundary. Backgrounded calls are meant to run within the turn (parallel to other steps) and be polled before completeJob aborts the controller; a non-request-scoped signal/connection that survives to enable true cross-turn MCP polling is the durable subagent follow-up. Note the new running-task reaper (f0a4dc906) also ensures an aborted/hung task is marked errored rather than hanging. Not re-fixing the lifecycle here by design.

File: packages/api/src/agents/handlers.ts [hunk 12]
```diff
@@ -3217,8 +3227,153 @@ export function createToolExecuteHandler(options: ToolExecuteOptions): EventHand
...
+                      )) as { content?: unknown; artifact?: unknown };
+                      /** Process artifacts (images, files, UI resources,
+                       *  citations) through the same callback as the foreground
+                       *  path so backgrounding an artifact-producing tool doesn't
+                       *  drop them. Best-effort: if the turn already returned and
+                       *  the stream is closed, persistence still runs and the SSE
+                       *  write no-ops. */
+                      if (toolEndCallback && result.artifact != null) {
+                        try {
+                          await toolEndCallback(
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/handlers.ts]: Re-raise of the decided after-save-artifact boundary. Within-turn completions attach best-effort; artifact-first built-ins are excluded; a persistence path not tied to the finalized turn's artifactPromises is the durable follow-up. Not re-fixing here by design.

**Comment by danny-avila** [comment]
**Codex round 8 addressed in `f0a4dc906`.** Two real registry fixes (resolved threads deleted):

- **Direct-poll TTL:** `get()` now sweeps before returning, so repeatedly polling a known `background_task_id` can't keep an expired completed task (and its retained result) alive past the one-hour TTL.
- **Stuck-task reaper:** `sweep()` now marks `running` tasks older than a 30-min running TTL as errored, so a detached call that never settles can't hold a running slot and exhaust the per-conversation cap.

The two intentional durable-boundary items (request/MCP-abort lifetime + after-save artifacts) were re-raised and remain the durable subagent follow-up — replied, not re-fixed. (The reaper does make an aborted/hung task report `error` cleanly rather than hang.)

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/handlers.ts [hunk 13]
```diff
@@ -3217,8 +3227,153 @@ export function createToolExecuteHandler(options: ToolExecuteOptions): EventHand
...
+                    backgroundUserId,
+                    backgroundConversationId,
+                    task.id,
+                    `Tool ${tc.name} not found`,
+                  );
+                } else {
+                  const strippedArgs = stripRunInBackgroundArg(tc.args);
+                  void (async () => {
+                    try {
+                      const result = (await tool.invoke(
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/handlers.ts]: Acknowledged, and noted that this is now P1. This is the decided same-run boundary: request-scoped MCP connections are torn down at response cleanup, so a backgrounded MCP call that OUTLIVES the turn can lose its connection. The feature is scoped to within-turn parallelism (dispatch, do other steps, poll before the turn ends); genuine cross-turn survival needs a background-owned connection/lifecycle and is the durable subagent follow-up. I'm surfacing the P1 escalation to the maintainer for the merge-vs-extend call rather than partially patching the lifecycle here.

File: packages/api/src/agents/handlers.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/handlers.ts]: Acknowledged (P1). Same decided boundary: completeJob aborts the run controller at turn end, so a task still running afterward is cancelled. Within-turn use is supported; a non-completing signal for detached work that survives the turn is the durable follow-up. Surfacing the escalation to the maintainer for the scope decision rather than patching here.

**Comment by danny-avila** [comment]
**Codex round 9 addressed in `66ecdf1f7`.** One real fix (resolved thread deleted):

- **Bucket-cap eviction:** only the running-task cap gates dispatch now; the total-tasks cap evicts the oldest settled tasks to make room instead of rejecting, so a burst of quick background calls in one conversation can't block new dispatches for up to the completed-TTL.

**Scope note:** Codex escalated the two boundary findings (request/MCP-connection lifetime + run-abort-signal detachment) to **P1** this round, arguing the feature 'mostly returns errors for long-running MCP calls' without them. That's a fair characterization of the same-run scope: this PR supports **within-turn** background parallelism (dispatch → do other steps → poll before the turn ends), and **cross-turn survival of long-running MCP calls is intentionally the durable subagent follow-up** (a background-owned connection + non-completing signal over the checkpointer/GenerationJobManager substrate). Flagging the P1 escalation to the maintainer for the merge-vs-extend decision rather than partially patching the request lifecycle here.

**Comment by danny-avila** [comment]
@codex review

**Comment by danny-avila** [comment]
**Contract reframed to within-turn in `6f36bec9f`** (directly addressing the two P1 lifecycle findings).

Rather than partially patch the request lifecycle, the model-facing contract now matches what the same-run implementation actually delivers. The `run_in_background` param description, `check_background_task` description, and the dispatch handle message all instruct the model to **collect the result within the same turn** (backgrounded work isn't guaranteed to survive past the turn end). So the feature is honestly scoped as **within-turn parallelism** — dispatch a slow call, do other steps, poll and collect before finishing the response — which the abort/connection lifecycle fully supports.

Cross-turn survival of genuinely long-running calls (a background-owned MCP connection + non-completing signal, or the subagent model over the checkpointer/GenerationJobManager substrate) remains the deliberate durable follow-up. The P1 threads are left open as trackers for that follow-up; the advertised contract no longer over-promises.

**Comment by danny-avila** [comment]
@codex review

**Comment by danny-avila** [comment]
@codex review

Extended background tool calls from within-turn to **cross-turn on a single process**, leak-free:

- The run's abort signal never reaches the detached invoke (the graph forwards only `configurable`/`metadata` to the tool-execute handler, never `signal`), so the floating promise survives turn completion; the in-process registry `get`/`list` key only on `user::conversation + id`, so a later turn polls the same task.
- **No connection leak:** ephemeral request-scoped MCP tools (runtime `{{LIBRECHAT_BODY_*}}` placeholders) capture their request store at creation and fall back to it, so config can't redirect them and their connection is torn down at request end. They are tagged in `createToolInstance` and run in the foreground instead of being backgrounded. Pooled/app-level MCP and structured tools are unaffected.
- Model-facing contract reworded from within-turn to cross-turn (not across restart/replica, which stays the durable follow-up).

New commit `99a51dc0c`. Tests added: cross-turn poll retrieval; ephemeral MCP tool runs foreground.

**Comment by danny-avila** [comment]
@codex review

Fixed in `d9d2865cd`: the ephemeral MCP tag now guards a null/stale `capturedServerConfig` (`capturedServerConfig ? requiresEphemeralUserConnection(...) : false`), matching the serverConfig-guarded callers elsewhere in the file. This also resolves the `Tests: api (shard 2/3)` failure (a null config threw via `isUserSourced`). 62 MCP unit tests pass locally.

**Comment by danny-avila** [comment]
@codex review

Fixed in `820088a12`: backgrounded-tool artifacts are no longer dropped after the turn. Instead of pushing to the finalized dispatch turn's `artifactPromises` (already awaited, stream closed), the artifact is held on the task and delivered through the **live poll turn's** `toolEndCallback` the first time `check_background_task` collects that id (once, then cleared to free memory), attributed to the original tool. Since the model must poll to collect any background result, same-turn and cross-turn share this path. Tests added: registry claim-once; artifact delivered on poll (not dispatch) and idempotent on re-poll.

**Comment by danny-avila** [comment]
@codex review

**Comment by danny-avila** [comment]
@codex review

**Comment by danny-avila** [comment]
@codex review

**Comment by danny-avila** [comment]
@codex review
